File: lwjgl/src/main/java/com/darkraha/opengldemoj/gl/Matrices.java
package com.darkraha.opengldemoj.gl;

import org.joml.Matrix4f;
import org.lwjgl.system.MemoryUtil;

import java.nio.FloatBuffer;

import static com.darkraha.opengldemoj.gl.GlUtils.*;

public class Matrices {

    public static final float PI_F = (float) Math.PI;
    public static final float TO_RAD = PI_F / 180.f;

    public final Matrix4f projection = new Matrix4f().perspective(45 * TO_RAD, 1, 1f, 100);
    public final Matrix4f camera = new Matrix4f().lookAlong(0f,0f,-1f,0,1,0);
    public Matrix4f model;
    public final Matrix4f viewModel = new Matrix4f();
    public final Matrix4f normals = new Matrix4f();
    public final Matrix4f matrix = new Matrix4f();



    public void applyModel(Matrix4f modelMatrix) {
        this.model = modelMatrix;
        camera.mul(modelMatrix, viewModel);
        matrix.set(projection).mul(viewModel);
        viewModel.invert(normals);
        normals.transpose();
    }


}

File: android/app/src/main/java/com/darkraha/gldemos/gl/Matrices.kt
package com.darkraha.gldemos.gl

import org.joml.Matrix4f

class Matrices {

    val projection = Matrix4f().perspective(45 * TO_RAD, 1f, 1f, 100f)
    val view = Matrix4f().lookAlong(0f, 0f, -1f, 0f, 1f, 0f)
    lateinit var model: Matrix4f
    val viewModel = Matrix4f()
    val normals = Matrix4f()
    val matrix = Matrix4f()

    fun applyModel(modelMatrix: Matrix4f) {
        model = modelMatrix
        view.mul(modelMatrix, viewModel)
        matrix.set(projection).mul(viewModel)
        viewModel.invert(normals)
        normals.transpose()
    }

    companion object {
        const val PI_F = Math.PI.toFloat()
        val TO_RAD: Float = PI_F / 180f
    }
}